prinz/src/main/scala/pl/touk/nussknacker/prinz/model/proxy/ProxiedInputModel.scala

```diff
@@ -17,33 +21,49 @@ class ProxiedInputModel(model: Model,
     originalModelInstance.signatureProvider
   ) {
     override def run(inputMap: VectorMultimap[String, AnyRef]): ModelRunResult = {
-      val modifiedInputMap = supplyNonProvidedInputs(inputMap)
-      originalModelInstance.run(modifiedInputMap)
+      val addInputParams = supplyNonProvidedInputs(inputMap)
+      val addComposedParams = addInputParams.flatMap(supplyNonProvidedComposedInputs)
+      addComposedParams.flatMap(originalModelInstance.run)
     }
   }
 
-  override def getName: ModelName = {
-    val originalName = model.getName.internal
-    new ModelName(s"Proxied-$originalName")
-  }
+  override def getName: ModelName = proxiedName
 
   override def getVersion: ModelVersion = model.getVersion
 
   override def toModelInstance: ModelInstance = proxiedModelInstance
 
-  private def supplyNonProvidedInputs(inputMap: VectorMultimap[String, AnyRef]): VectorMultimap[String, AnyRef] =
-    proxiedParams
-      .filter(inputsToBeReplacedIn(inputMap))
-      .map(supplyParamValue)
-      .foldLeft(inputMap) { (acc, value) => acc.add(value._1, value._2) }
+  private def supplyNonProvidedInputs(inputMap: VectorMultimap[String, AnyRef]): Future[VectorMultimap[String, AnyRef]] =
+    Future.sequence(
+      proxiedParams
+        .filter(inputsToBeReplacedIn(inputMap))
+        .map(supplyParamValue)
+    ).map(addExtraInputsTo(inputMap))
 
-  private def supplyParamValue(param: ProxiedModelInputParam): (String, AnyRef) = {
+  private def supplyNonProvidedComposedInputs(inputMap: VectorMultimap[String, AnyRef]): Future[VectorMultimap[String, AnyRef]] =
+    Future.sequence(composedProxiedParams.map { param => supplyComposedParamValues(param) } )
+      .map(_.foldLeft(inputMap) { (acc, composedValues) =>
+        addExtraInputsTo(acc)(composedValues)
+      })
+
+  private def supplyParamValue(param: ProxiedModelInputParam): Future[(String, AnyRef)] = {
     val paramName = param.paramName
     val metadata = ModelInputParamMetadata(paramName, modelMetadata)
-    val result = param.paramSupplier(metadata)
-    (paramName.name, result)
+    val resultFuture = param.paramSupplier(metadata)
+    resultFuture.map { result =>
+      (paramName.name, result)
+    }
   }
 
   private def inputsToBeReplacedIn(inputMap: VectorMultimap[String, AnyRef]): ProxiedModelInputParam => Boolean =
-    param => !inputMap.containsKey(param.paramName.name) || param.overwriteProvided
+    param => !inputMap.containsKey(param.paramName.name)
+
+  private def supplyComposedParamValues[T <: AnyRef](param: ProxiedModelComposedInputParam[T]): Future[Iterable[(String, AnyRef)]] = {
+    param.paramsSupplier(modelMetadata)
+      .flatMap(param.paramsExtractor(_))
```

Comment:
this can be in one method in ProxiedModelComposedInputParam?

Reply:
Done